### PR TITLE
RavenDB-19560 Can store and project dictionary from Index. 

### PR DESCRIPTION
**Pull request by maciejaszyk**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19560

### Additional description

IDictionary was detected as `IEnumerable` and serialized as `ValueKeyPair<T,T>`.  

I've introduced a new type that match `IDictionary` in `LuceneDocumentQueryBase`. Since version 6.0 we're gonna save `IDictionary` as a JSON.

### Type of change

- Bug fix

### How risky is the change?

- Moderate 

### Backward compatibility
- Breaking change

### Is it platform specific issue?

- No

### Documentation update

- This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/Indexes/Persistence/ConverterBase.cs [hunk 1]
```diff
@@ -109,6 +109,9 @@ protected static ValueType GetValueType(object value)
             if (value is DynamicDictionary)
                 return ValueType.ConvertToJson;
 
+            if (value is IDictionary && _index.Definition.Version >= IndexDefinitionBaseServerSide.IndexVersion.ProperlyParseDictionaryToStoredField)
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Persistence/ConverterBase.cs:112]: Also need to handle `IDictionary<string,string>`, and a few other permutation, no?

Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Persistence/ConverterBase.cs:112]: https://dotnetfiddle.net/FrjRuL

It works for `IDictionary<string,string>`. I cannot figure out the case it doesn't.

Comment (ayende) [reply src/Raven.Server/Documents/Indexes/Persistence/ConverterBase.cs:112]: Thanks, did *not* know that.

File: src/Raven.Server/Documents/Indexes/IndexDefinitionBaseServerSide.cs [hunk 2]
```diff
@@ -150,10 +150,13 @@ public static class IndexVersion
             public const long ProperlyParseThreeDigitsMillisecondsDates = 52_002; // RavenDB-17711
 
             public const long EngineTypeStored = 54_000; // introducing Corax, added engine type to the index storage
+            
+            public const long ProperlyParseDictionaryToStoredField = 60_000; // RavenDB-17711
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/IndexDefinitionBaseServerSide.cs]: The comment mentioning `RavenDB-17711` is not the valid issue number

**Review by arekpalinski — APPROVED** [review]

File: src/Raven.Server/Documents/Indexes/Persistence/ConverterBase.cs [hunk 3]
```diff
@@ -264,6 +267,8 @@ protected enum ValueType
             CoraxSpatialPointEntry,
             
             CoraxDynamicItem,
+            
+            IDictionary
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Indexes/Persistence/ConverterBase.cs]: just 'Dictionary`

**Review by ppekrol — APPROVED** [review]

**Pull request merged**